tests/tst_ncio_mpi.F90

```diff
@@ -0,0 +1,198 @@
+! This is a test for the NCEPLIBS-ncio package.
+! Author: Brian Curtis (brian.curtis@noaa.gov) May 2021
+program tst_ncio_mpi
+
+  use mpi
+  use netcdf
+  use module_ncio
+  implicit none
+
+  character(len=72) charatt, time_units
+  type(Dataset) :: dset, dsetin
+  type(Variable) :: var
+  real(4), allocatable, dimension(:) :: values_1d
+  real(4), allocatable, dimension(:,:) :: values_2d
+  real(4), allocatable, dimension(:,:,:) :: values_3d
+  real(4), allocatable, dimension(:,:,:,:) :: values_4d
+  real(4), allocatable, dimension(:,:,:,:,:) :: values_5d
+  real(4), dimension(10,10) :: quantize1, quantize2
+  real(4) mval,r4val,qerr
+  integer ndim,nvar,ndims,ival,idate(6),icheck(6),ierr,n,nbits
+  integer :: my_rank, nprocs
+  integer :: mpi_err
+  logical hasit
+
+  call mpi_init(mpi_err)
+  call MPI_Comm_rank(MPI_COMM_WORLD, my_rank, mpi_err)
+  call MPI_Comm_size(MPI_COMM_WORLD, nprocs, mpi_err)
+
+  if (my_rank .eq. 0) print*, '*** Testing NCEPLIBS-ncio with MPI.'
+  dsetin = open_dataset('dynf000_par_template.nc.in', paropen=.true.)
+
+  if (my_rank .eq. 0) print *,'*** Test creation of new dataset from template...'
+  dset = create_dataset('dynf000_par.nc', dsetin, paropen=.true.)
+
+  if (my_rank .eq. 0) print *,'*** Test that number of variables,dimensions,attributes is read...'
+  if (dsetin%nvars .ne. 24) stop 4
+
+  if (dsetin%ndims .ne. 6) stop 5
+
+  if (dsetin%natts .ne. 8) stop 6
+
+
+  if (my_rank .eq. 0) print *,'*** Test read of variable data...'
+  call read_vardata(dsetin, 'pressfc', values_3d)
+  call read_vardata(dsetin, 'vgrd', values_4d)
+  call read_vardata(dsetin, 'tmp_spread', values_5d)
+
+  if (maxval(values_3d) .ne. 102345.6) stop 7
+
+  if (minval(values_4d) .ne. -5.5) stop 8
+  if ((minval(values_5d) .ne. -1.0) .and. (maxval(values_5d) .ne. 1.0)) stop 9
+
+  call close_dataset(dsetin)
+  values_3d=1.013e5
+  values_4d=99.
+
+  ! populate pressfc and vgrd
+  if (my_rank .eq. 0) print *,'*** Test write of variable data...'
+  !!call write_vardata(dset,'pressfc', values_3d)
+  !! SKIP THESE TWO
+  call write_vardata(dset,'vgrd', values_4d)
+  !!
+  !! ! write just a slice along 3rd dimension (index 10)
+  !!values_3d = -99.
+  !!call write_vardata(dset, 'vgrd', values_3d, 10, 3)
```

Comment:
@BrianCurtis-NOAA is this the call that hangs on MacOS?  This is not how parallel IO is supposed to work. Each processor needs to figure out which 'domain' it is to read/write for it to be effectively tested. This seems like each PE is trying to write the same exact grid points which could be the trouble.

Reply:
@CoryMartin-NOAA OK, I think I understand better. Does this make sense: Currently I am just reading all data for each rank (processor) and to test MPI I need to have the data divided in such a way that each processor writes without overlapping data points. (in a better MPI test, i would have MPI read in this fashion as well?)

I'm not sure how to properly divide the data. Is this a simple explanation or is there a resource that may help me?

Reply:
There are different ways to divide the data to read in parallel. One simple way for a 3D array is to read only specific 'layers' on each MPI task. In this ncio wrapper module, there are variables, `ncstart` and `nccount`. These just need to be computed as a function of MPI task number. If your array has dimensions 10,10,4, for example, and you use 4 MPI tasks, then you can read 10,10,1 on each task (that is the count) and the start would be 1,1,mype+1.

Also note that after read or before write, you will want to do something like MPI gather/scatter to distribute the data properly.